lib/buffered-runner.js

```diff
@@ -0,0 +1,247 @@
+'use strict';
+
+const allSettled = require('promise.allsettled');
+const Runner = require('./runner');
+const {EVENT_RUN_BEGIN, EVENT_RUN_END} = Runner.constants;
+const debug = require('debug')('mocha:parallel:buffered-runner');
+const {WorkerPool} = require('./pool');
+const {setInterval, clearInterval} = global;
+const {createMap} = require('./utils');
+
+/**
+ * Outputs a debug statement with worker stats
+ * @param {WorkerPool} pool - Worker pool
+ */
+const debugStats = pool => {
+  const {totalWorkers, busyWorkers, idleWorkers, pendingTasks} = pool.stats();
+  debug(
+    '%d/%d busy workers; %d idle; %d tasks queued',
+    busyWorkers,
+    totalWorkers,
+    idleWorkers,
+    pendingTasks
+  );
+};
+
+/**
+ * The interval at which we will display stats for worker processes in debug mode
+ */
+const DEBUG_STATS_INTERVAL = 5000;
+
+const ABORTED = 'ABORTED';
+const IDLE = 'IDLE';
+const ABORTING = 'ABORTING';
+const RUNNING = 'RUNNING';
+const BAILING = 'BAILING';
+const BAILED = 'BAILED';
+const COMPLETE = 'COMPLETE';
+
+const states = createMap({
+  [IDLE]: new Set([RUNNING, ABORTING]),
+  [RUNNING]: new Set([COMPLETE, BAILING, ABORTING]),
+  [COMPLETE]: new Set(),
+  [ABORTED]: new Set(),
+  [ABORTING]: new Set([ABORTED]),
+  [BAILING]: new Set([BAILED, ABORTING]),
+  [BAILED]: new Set([COMPLETE, ABORTING])
+});
+
+/**
+ * This `Runner` delegates tests runs to worker threads.  Does not execute any
+ * {@link Runnable}s by itself!
+ */
+class BufferedRunner extends Runner {
```

Comment:
> Why is it called a "Buffered" runner? Since the description isn't explaining that either. The fact that it uses a buffer internally doesn't help the consumer of this class.

`BufferedParallelRunner` could be more accurate.  The idea is that `ParallelRunner` is fine until we add an _unbuffered_ parallel runner; e.g., a streaming parallel runner.   This would be especially useful for streaming machine-readable results; not so much for human-readable output.  It would more closely mimic serial mode; how reporters generate output in real-time.  But it would also be confusing to use with a structured reporter like `spec`, because it'd just be a firehose with random indentation.  Would work for `dot`, though.

> Also: we're allowed to use es classes? Awesome! 😍

This is fair game in `bin/`, `scripts/`, `lib/nodejs/`, `lib/cli/` and `test/node-unit`... essentially, code that only runs in Node.js.

Reply:
> `BufferedParallelRunner` could be more accurate. The idea is that `ParallelRunner` is fine until we add an unbuffered parallel runner; e.g., a streaming parallel runner.

I would expect the `streaming` might be a _setting_ of the `ParellelRunner` in the future. It is a minor implementation detail, right? Just a `streaming` reporter configured in the worker process, rather than a `buffered` reporter.